packages/aws_signature_v4/test/c_test_suite/request_parser.dart

```diff
@@ -0,0 +1,193 @@
+          break;
+        }
+        value += '\n' + line;
+        nextLineNo++;
+      }
+
+      // Concatenate values for the same key with a ','
+      headers.update(
+        key,
+        (currValue) => currValue + ',' + value,
+        ifAbsent: () => value,
+      );
+
+      // Skip to the next key line
+      lineNo = headerLines.indexWhere((el) => el.contains(':'), nextLineNo);
+    }
+
+    return headers;
+  }
+
+  /// Writes the raw request to the HTTP server and listens for the server to
+  /// emit the parsed [HttpRequest].
+  Future<AWSBaseHttpRequest> parse(
+    String request, {
+    required Context context,
+  }) async {
+    // Some request files have paths with space characters. The Dart server will
+    // not accept these and they must be percent-encoded.
+    request = _preprocessRequest(request);
+
+    // Use a different connection each time so don't need to bother sending
+    // a response from the server.
+    _sendSocket = await Socket.connect(
+      InternetAddress.loopbackIPv4,
+      _server.port,
+    );
+
+    // Send request and wait for server to process it.
+    final nextRequest = _serverStream.first;
+    _sendSocket.write(request);
+    _sendSocket.writeln();
+    await _sendSocket.flush();
+    await _sendSocket.close();
+    final HttpRequest httpRequest = await nextRequest;
+
+    // Capture the body bytes of the request.
+    final List<int> body = await ByteStream(httpRequest).toBytes();
+
+    // Close server connection.
+    await httpRequest.response.close();
+
+    final requestLines = request.split(RegExp('\r?\n'));
+
+    // Parse these parameters manually since they are modified, normalized, etc.
+    // by the Dart server.
+    final path = _parseRequestPath(requestLines.first);
+    final queryParameters = _parseQueryString(requestLines.first);
+    final headers = _parseHeaders(requestLines.sublist(1));
+    final host = headers[AWSHeaders.host]!;
+
+    return AWSHttpRequest.raw(
+      method: AWSHttpMethodHelper.fromString(httpRequest.method),
+      host: host,
+      path: path,
+      headers: headers,
+      queryParameters: queryParameters,
+      body: body,
+    );
+  }
+
+  /// Encodes spaces in the URI before sending to the Dart server. This might
+  /// be an implementation detail of the server but it cannot process the request
+  /// otherwise.
+  String _preprocessRequest(String request) {
+    final requestLines = request.split(RegExp('\r?\n'));
+    var requestLine = requestLines.first;
+
+    final requestParts = requestLine.split(' ');
+    final requestPartsLength = requestParts.length;
+
+    // Only perform work if the request path has spaces.
+    //
+    // e.g. requestLine = 'GET /abc 123/ HTTP/1.1'
+    //   => requestParts = ['GET', '/abc', '123/', 'HTTP/1.1']
+    if (requestPartsLength > 3) {
+      requestLine = [
+        requestParts.first,
+        requestParts.sublist(1, requestPartsLength - 1).join('%20'),
```

Comment:
nit: would be nice to have a literal name for `'%20'` for readability, and also for `' '`